src/clone_area_cmd.cpp

```diff
@@ -0,0 +1,561 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file clone_area_cmd.cpp Commands related to clone area. */
+
+#include "stdafx.h"
+#include "command_func.h"
+#include "viewport_func.h"
+#include "company_base.h"
+#include "clone_area_cmd.h"
+#include "newgrf_station.h"
+#include "safeguards.h"
+#include "road.h"
+#include "road_internal.h"
+#include "depot_base.h"
+#include "tunnelbridge_map.h"
+#include "sound_func.h"
+#include "rail_cmd.h"
+#include "tunnelbridge_cmd.h"
+#include "station_cmd.h"
+#include "terraform_cmd.h"
+
+TileIndex selected_tile;
+TileIndex selected_start_tile;
+bool selected_diagonal;
+
+/**
+ * Rotate by an angle using the formula for rotating a point on a plane.
+ *
+ * @param point Rotating point.
+ * @param angle Rotate an angle.
+ */
+TileIndexDiffC Rotate(TileIndexDiffC point, DiagDirDiff angle) {
+	TileIndexDiffC rotated;
+	switch (angle) {
+		case DIAGDIRDIFF_90LEFT:
+			rotated.x = - point.y;
+			rotated.y = point.x;
+			break;
+		case DIAGDIRDIFF_REVERSE:
+			rotated.x = - point.x;
+			rotated.y = - point.y;
+			break;
+		case DIAGDIRDIFF_90RIGHT:
+			rotated.x = point.y;
+			rotated.y = - point.x;
+			break;
+		default:
+			return point;
+	}
+	return rotated;
```

Comment:
```suggestion
switch (angle) {
    case DIAGDIRDIFF_90LEFT: return {-point.y, point.x};
    case DIAGDIRDIFF_REVERSE: return {-point.x, -point.y};
    case DIAGDIRDIFF_90RIGHT: return {point.y, -point.x};
    default: return point;
}
```

Reply:
It wasn't a good idea
Warning: narrowing conversion of '-(int)point.TileIndexDiffC::y' from 'int' to 'int16_t' {aka 'short int'} [-Wnarrowing]